### PR TITLE
gh-106078: Convert `_decimal` types to heap types

### PR DESCRIPTION
**Pull request by CharlieZhao95**
* Establish a global module state
* Convert `_decimal` types to heap types

Co-authored-by: Erlend E. Aasland erlend.aasland@protonmail.com

* Issue: gh-106078

### CODE REVIEW

**Comment by CharlieZhao95** [comment]
cc: @kumaraditya303 @erlend-aasland

File: Misc/NEWS.d/next/Library/2023-06-25-16-54-24.gh-issue-106078.cFnYUM.rst
Comment (kumaraditya303) [inline Misc/NEWS.d/next/Library/2023-06-25-16-54-24.gh-issue-106078.cFnYUM.rst]: Maybe we can skip news for individual PRs and add one in the end?

File: Modules/_decimal/_decimal.c
Comment (kumaraditya303) [inline Modules/_decimal/_decimal.c]: Guard with `Py_DEBUG`

File: Modules/_decimal/_decimal.c
Comment (kumaraditya303) [inline Modules/_decimal/_decimal.c]: Guard with `Py_DEBUG`

File: Misc/NEWS.d/next/Library/2023-06-25-16-54-24.gh-issue-106078.cFnYUM.rst
Comment (CharlieZhao95) [reply Misc/NEWS.d/next/Library/2023-06-25-16-54-24.gh-issue-106078.cFnYUM.rst]: Yes. I agree with you. I checked the PRs related to isolate `_io` and that's what they do.

**Review by kumaraditya303 — COMMENTED** [review]
Mostly LGTM but there are some compiler warnings to be taken care of.

File: Misc/NEWS.d/next/Library/2023-06-25-16-54-24.gh-issue-106078.cFnYUM.rst
Comment (erlend-aasland) [reply Misc/NEWS.d/next/Library/2023-06-25-16-54-24.gh-issue-106078.cFnYUM.rst]: > Maybe we can skip news for individual PRs and add one in the end?

Yes, that's been the practice so far.

File: Modules/_decimal/_decimal.c
Comment (erlend-aasland) [inline Modules/_decimal/_decimal.c]:
```suggestion
#define PyDec_CheckExact(st, v) Py_IS_TYPE(v, (st)->PyDec_Type)
#define PyDec_Check(st, v) PyObject_TypeCheck(v, (st)->PyDec_Type)
#define PyDecSignalDict_Check(st, v) Py_IS_TYPE(v, (st)->PyDecSignalDict_Type)
#define PyDecContext_Check(st, v) PyObject_TypeCheck(v, st->PyDecContext_Type)
```

File: Modules/_decimal/_decimal.c
Comment (erlend-aasland) [inline Modules/_decimal/_decimal.c]: I'd use `modstate` or another more explicit name for this particular scope; the difference between `tstate` and `state` is small; it is easy to write or interpret it wrong. Explicit is nice.

File: Modules/_decimal/_decimal.c
Comment (erlend-aasland) [inline Modules/_decimal/_decimal.c:1812]: Can you also move the declaration of `self` down here?

File: Modules/_decimal/_decimal.c
Comment (erlend-aasland) [inline Modules/_decimal/_decimal.c]: Please follow PEP-7.

```suggestion
ctxmanager_traverse(PyDecContextManagerObject *self, visitproc visit,
                    void *arg)
```

File: Modules/_decimal/_decimal.c
Comment (erlend-aasland) [inline Modules/_decimal/_decimal.c]: Always use good macro hygiene :)

```suggestion
#define dec_alloc(st) PyDecType_New((st)->PyDec_Type)
```

File: Modules/_decimal/_decimal.c
Comment (erlend-aasland) [inline Modules/_decimal/_decimal.c]: Macro hygiene: please add parentheses to the state pointer chases here.

File: Modules/_decimal/_decimal.c
Comment (erlend-aasland) [inline Modules/_decimal/_decimal.c]: Why not just above line 3621 where it is used first?

File: Modules/_decimal/_decimal.c
Comment (erlend-aasland) [inline Modules/_decimal/_decimal.c]: Nitpick; I prefer the condensed variant:

```suggestion
#define CREATE_TYPE(MOD, TP, SPEC) do {                               \
    TP = (PyTypeObject *)PyType_FromMetaclass(NULL, MOD, SPEC, NULL); \
    CHECK_PTR(TP);                                                    \
} while (0)
```

File: Modules/_decimal/_decimal.c
Comment (erlend-aasland) [reply Modules/_decimal/_decimal.c]: Is state implicitly needed by the `CURRENT_CONTEXT` macro? If so, that's a bad design; if it is needed, it should be passed as a param, IMO.

File: Modules/_decimal/_decimal.c
Comment (CharlieZhao95) [reply Modules/_decimal/_decimal.c]: Looks good!

But macro parameters are usually named in the same style as function parameters, e.g. unix like style and camelCase, IMO. This helps us distinguish macro parameters from other macros or constants such as `NULL`.

Do you mind if I modify the style of macro parameters. :)

Comment (erlend-aasland) [reply Modules/_decimal/_decimal.c]: Well, it's a personal preference ;) You'll find both styles in the CPython code base. I've come to like the upper case spelling for the following reasons:

- I find it easier to read (personal preference)
- Most variable names are lower case; using upper case inside the macro prevents you from accidentally shadowing variables in your surrounding scope

I said my preference; do as you wish ;)

File: Modules/_decimal/_decimal.c
Comment (CharlieZhao95) [reply Modules/_decimal/_decimal.c]: > Is state implicitly needed by the `CURRENT_CONTEXT` macro? If so, that's a bad design; if it is needed, it should be passed as a param, IMO.

**`CURRENT_CONTEXT` macro doesn't implicitly depend on `state`.** 

The `current_context` function contains global variable that need to be moved to the module state. In the [original PR](https://github.com/python/cpython/pull/103381), I pass the `state` as a parameter to the `CURRENT_CONTEXT` macro. When I broke the PR, I did not adjust the position of the statement. That's why the statement is above the `CURRENT_CONTEXT` macro. :)

I will change the position of this statement to avoid misleading.

Comment (erlend-aasland) [reply Modules/_decimal/_decimal.c]: Thanks; there's a lot of confusing occurrences like this in this PR. Please take care of the other occurrences as well.

**Comment by erlend-aasland** [comment]
Can you please add tests for basic type qualities:

- immutability
- disallow instantiation

**Comment by kumaraditya303** [comment]
Refleaks looks good:
```console
0:00:00 load avg: 2.73 Run tests sequentially
0:00:00 load avg: 2.73 [1/1] test_decimal
beginning 6 repetitions
123456
......

== Tests result: SUCCESS ==

1 test OK.

Total duration: 26.0 sec
Tests result: SUCCESS
```

File: Lib/test/test_decimal.py
Comment (erlend-aasland) [inline Lib/test/test_decimal.py]: Wow, **decimal** must be the only remaining stdlib module with such an ancient test setup. I remember Serhiy cleaning up and modernising the stdlib test suite a couple of years ago; I guess something blocked that change from happening for the decimal module.

Comment (erlend-aasland) [reply Lib/test/test_decimal.py]: cc. @serhiy-storchaka, do you remember the details?

**Comment by CharlieZhao95** [comment]
I found another interesting problem while testing. The following code will causes a **segmentation fault**:

```python
>>> import decimal
>>> tp = type(decimal.Context().flags)  # SignalDict type
>>> tp()  # Segmentation fault
```

I guess that the problem is caused by accessing the wild pointer when executing the `signaldict_repr` function. Maybe this problem is worth to open a new issue.

File: Lib/test/test_decimal.py
Comment (erlend-aasland) [reply Lib/test/test_decimal.py]: IIRC, this is the issue: gh-89392

Comment (erlend-aasland) [reply Lib/test/test_decimal.py]: Ah, see https://github.com/python/cpython/issues/89392#issuecomment-1128465649

**Review by serhiy-storchaka — COMMENTED** [review]
There are currently two test classes for the C implementation: CWhitebox and SignatureTest. None of them are suitable for new tests?

**Comment by erlend-aasland** [comment]
> There are currently two test classes for the C implementation: CWhitebox and SignatureTest. None of them are suitable for new tests?

Good point; I think the added tests can fit in the CWhitebox test case.

File: Modules/_decimal/_decimal.c
Comment (kumaraditya303) [inline Modules/_decimal/_decimal.c]:
```suggestion
#define PyDecContext_Check(st, v) PyObject_TypeCheck(v, (st)->PyDecContext_Type)
```

**Review by kumaraditya303 — APPROVED** [review]

**Comment by kumaraditya303** [comment]
I'll merge this later today after a refleak test.

**Comment by erlend-aasland** [comment]
> I'll merge this later today after a refleak test.

@kumaraditya303, let me amend the tests before merging.

**Comment by serhiy-storchaka** [comment]
It would be nice to wait until #79654 be merged.

**Review by erlend-aasland — APPROVED** [review]

**Comment by erlend-aasland** [comment]
Good job, @CharlieZhao95!

**Comment by CharlieZhao95** [comment]
> Good job, @CharlieZhao95!

Thanks all! Let's move on to the next! :)

**Pull request merged**
